go/enclave/rollupchain/rollup_chain.go

```diff
@@ -387,78 +339,119 @@ func (rc *RollupChain) insertBlockIntoL1Chain(block *types.Block, isLatest bool)
 
 // Updates the L1 and L2 chain heads, and returns the new L2 head hash and the produced batch, if there is one.
 func (rc *RollupChain) updateL1AndL2Heads(block *types.Block) (*common.L2RootHash, *core.Batch, error) {
-	// We extract the rollups from the block.
-	rollupsInBlock := rc.bridge.ExtractRollups(block, rc.storage)
-
-	// We determine whether this is a pre-genesis, genesis, or post-genesis block, and handle the block.
-	blockStage, err := rc.getBlockStage(rollupsInBlock)
+	err := rc.processRollups(block)
 	if err != nil {
-		return nil, nil, fmt.Errorf("could not determine block type. Cause: %w", err)
+		return nil, nil, fmt.Errorf("could not process rollup in block. Cause: %w", err)
 	}
 
-	switch blockStage {
-	case PreGenesis:
-		return nil, nil, nil
-	case Genesis:
-		l2Head, err := rc.handleGenesisBlock(block, rollupsInBlock)
-		return l2Head, nil, err
-	case PostGenesis:
-		return rc.handlePostGenesisBlock(block, rollupsInBlock)
-	}
-	return nil, nil, fmt.Errorf("unrecognised block stage")
-}
-
-// Determines if this is a pre-genesis L2 block, the genesis L2 block, or a post-genesis L2 block.
-func (rc *RollupChain) getBlockStage(rollupsInBlock []*core.Rollup) (BlockStage, error) {
-	_, err := rc.storage.FetchBatchByHeight(0)
+	// We determine whether this is the genesis block.
+	isGenesis := false
+	l2Head, err := rc.storage.FetchHeadBatchForBlock(block.ParentHash())
 	if err != nil {
 		if !errors.Is(err, errutil.ErrNotFound) {
-			return -1, fmt.Errorf("could not retrieve genesis rollup. Cause: %w", err)
+			return nil, nil, fmt.Errorf("could not retrieve current head batch. Cause: %w", err)
+		}
+		isGenesis = true
+	}
+
+	var l2HeadTxReceipts types.Receipts
+	// If we're the sequencer, we produce a new L2 head to replace the old one.
+	if rc.nodeType == common.Sequencer {
+		l2Head, l2HeadTxReceipts, err = rc.produceAndStoreBatch(block, isGenesis)
```

Comment:
I'm a bit worried that `isLatest` doesn't seem to be a factor in whether we produce a batch, is that correct?

As far as I can tell, it won't error during the `updateL1State` and so it'll fall all the way through to here, and produce one. So if we're behind the L1 we could be potentially be spamming out batches with transactions based on old state.

I'm a bit fuzzy on how this should work tbh but there's got to be some sort of alignment going on.